features/mbedtls/targets/TARGET_Cypress/TARGET_PSOC6/aes_alt.c

```diff
@@ -0,0 +1,747 @@
+    cy_en_crypto_status_t status;
+
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( ctx->aes_state.buffers != NULL );
+
+    status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_ENCRYPT, output, input, &ctx->aes_state);
+
+    if (CY_CRYPTO_SUCCESS != status)
+    {
+    	ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+    }
+
+    return( ret );
+}
+
+#if !defined(MBEDTLS_DEPRECATED_REMOVED)
+void mbedtls_aes_encrypt( mbedtls_aes_context *ctx,
+                          const unsigned char input[16],
+                          unsigned char output[16] )
+{
+    mbedtls_internal_aes_encrypt( ctx, input, output );
+}
+#endif /* !MBEDTLS_DEPRECATED_REMOVED */
+
+/*
+ * AES-ECB block decryption
+ */
+int mbedtls_internal_aes_decrypt( mbedtls_aes_context *ctx,
+                                  const unsigned char input[16],
+                                  unsigned char output[16] )
+{
+    int ret = 0;
+    cy_en_crypto_status_t status;
+
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( ctx->aes_state.buffers != NULL );
+
+    status = Cy_Crypto_Core_Aes_Ecb(ctx->obj.base, CY_CRYPTO_DECRYPT, output, input, &ctx->aes_state);
+
+    if (CY_CRYPTO_SUCCESS != status)
+    {
+    	ret = MBEDTLS_ERR_AES_HW_ACCEL_FAILED;
+    }
+
+    return( ret );
+}
+
+#if !defined(MBEDTLS_DEPRECATED_REMOVED)
+void mbedtls_aes_decrypt( mbedtls_aes_context *ctx,
+                          const unsigned char input[16],
+                          unsigned char output[16] )
+{
+    mbedtls_internal_aes_decrypt( ctx, input, output );
+}
+#endif /* !MBEDTLS_DEPRECATED_REMOVED */
+
+/*
+ * AES-ECB block encryption/decryption
+ */
+int mbedtls_aes_crypt_ecb( mbedtls_aes_context *ctx,
+                    int mode,
+                    const unsigned char input[16],
+                    unsigned char output[16] )
+{
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+
+    if( mode == MBEDTLS_AES_ENCRYPT )
+        return( mbedtls_internal_aes_encrypt( ctx, input, output ) );
+    else
+        return( mbedtls_internal_aes_decrypt( ctx, input, output ) );
+}
+
+#if defined(MBEDTLS_CIPHER_MODE_CBC)
+/*
+ * AES-CBC buffer encryption/decryption
+ */
+int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
+                    int mode,
+                    size_t length,
+                    unsigned char iv[16],
+                    const unsigned char *input,
+                    unsigned char *output )
+{
+    unsigned char temp[CY_CRYPTO_AES_BLOCK_SIZE];
+
+    int ret = 0;
+    cy_en_crypto_status_t status;
+
+    AES_VALIDATE_RET( ctx != NULL );
+    AES_VALIDATE_RET( mode == MBEDTLS_AES_ENCRYPT ||
+                      mode == MBEDTLS_AES_DECRYPT );
+    AES_VALIDATE_RET( iv != NULL );
+    AES_VALIDATE_RET( input != NULL );
+    AES_VALIDATE_RET( output != NULL );
```

Comment:
Is the input buffer being non-NULL the only requirement for the AES accelerator to work? Are there also memory alignment or size requirements? Any bus requirements? We've seen accelerators that require 16 byte aligned addresses, only work from RAM (can't use flash directly), or only work with `length` as a multiple of the block size.

Reply:
Short answer is No. This Crypto HW block have master interface on AHB bus with 8/16/32-bit accesses to memory space in PSoC (including RAM and Flash)

Reply:
ALT implementation uses pdl crypto driver functionality that doesn't have any bus or alignment requirements. PDL crypto driver has own internal buffers and it's manipulation code

Reply:
OK, sounds good